frontend/catalyst/api_extensions/callbacks.py

```diff
@@ -39,6 +40,95 @@
 from catalyst.utils.types import convert_pytype_to_shaped_array
 
 
+class CallbackWithCustomGrad:
+    """A callback with a custom grad"""
+
+    def __init__(self, func, restype, forward, reverse):
+        assert func and forward and reverse
+        self.func = func
+        self.restype = restype
+        self._fwd = forward
+        self._fwd_jaxpr = None
+        self._bwd = reverse
+        self._bwd_jaxpr = None
+        self.callback = None
+
+    def __call__(self, *args, **kwargs):
+        if self.callback:
+            return self.callback(*args, **kwargs)
+
+        def closure(*args, **kwargs) -> self.restype:
+            return self.func(*args, **kwargs)
+
+        # We need this here to avoid infinite recursion
+        # Where does the infinite recursion happen?
+        # It happens if the fwd or bwd passes have a call to
+        # the pure_callback implementation.
+        self.callback = base_callback(closure, custom_grad=self)
+
+        # The arguments here are tracers.
+        # And we want to just get the abstraction of the tracers (i.e., the types)
+        absargs, abskwargs = tree_map(shaped_abstractify, (args, kwargs))
+        # Once we have the types, we can call this self.func with the absargs and abskwargs.
+        # We don't need the jaxpr representation but the output is the shape of the cotangents.
+        _jaxpr, cotangents = jax.make_jaxpr(self, return_shape=True)(*absargs, **abskwargs)
```

Comment:
Are the cotangents not just going to be `self.restype` always?

Reply:
https://github.com/PennyLaneAI/catalyst/pull/834/commits/c7b80d3fce9f7b5c8a08b1325a4fdcdac2407120